Added validation and error checking for incoming updates in ListenFor…

dmitriy-kharchenko opened the pull request:
if you open the URL of the webhook in the browser for example to debug... then your app will crash due to memory violation.
These Changes fix it!:)

AstolfoKawaii commented:
LGTM, but maybe, you should use `MethodPost` instead of the "POST", but it's just a meter of taste, I guess...
May I include this in my own fork?
Thank you.

dmitriy-kharchenko commented:
Hello, @AstolfoKawaii.
Yes, sure :)
I changed "POST" to `MethodPost`

Syfaro commented:
Looks like a good improvement! I let it sit too long though, any chance you could resolve the merge conflicts?

dmitriy-kharchenko commented:
Hello @Syfaro :) I have updated and refactored code but tests were failed.

```
=== RUN   TestSendWithContact
2020/07/21 09:08:17 sendContact resp: {"ok":false,"error_code":429,"description":"Too Many Requests: retry after 34433","parameters":{"retry_after":34433}}
--- FAIL: TestSendWithContact (0.22s)
	bot_test.go:261: Too Many Requests: retry after 34433
```

Could you suggest any ideas where search error/bug?

AstolfoKawaii commented:
> Could you suggest any ideas where search error/bug?

This test fails in the upstream either, so that's not your bug. https://travis-ci.org/github/go-telegram-bot-api/telegram-bot-api/builds/710295675

Syfaro commented:
@dmitriy-kharchenko your code is likely fine, it's Telegram rate limiting interfering with testing. 

Thanks for updating the PR so quickly, I'm going to give it some local testing but it looks good so far!

tjhorner commented on bot.go:
```diff
@@ -557,21 +557,44 @@ func (bot *BotAPI) ListenForWebhook(pattern string) UpdatesChannel {
 
-  var update Update
-  json.Unmarshal(bytes, &update)
+	payload, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := r.Body.Close(); err != nil {
+		return nil, err
+	}
+
+	var update Update
+	err = json.Unmarshal(payload, &update)
+	if err != nil {
+		return nil, err
+	}
```
While we are here, how about we use `json.Decoder` instead of `json.Unmarshal`? This way we don't need to load the entire body into memory first. Well, loading it into memory is [essentially what Go does internally](https://github.com/golang/go/blob/85afa2eb190d5d1a06584803bde4b4ee9b0e79b0/src/encoding/json/stream.go#L49-L79), but I think it is better to let the language worry about the implementation of decoding a JSON stream. (Plus, it has some other niceties like failing early if a token is not allowed as it's streamed in, closing the body stream for us, etc.)

Could do something like:

```suggestion
	var update Update
	err := json.NewDecoder(r.Body).Decode(&update)
	if err != nil {
		return nil, err
	}
```

dmitriy-kharchenko commented:
@tjhorner Thank you very much for this improvement!)
